cli/internal/dataplane/write.go

```diff
@@ -172,11 +181,27 @@ func Write(ctx context.Context, uri *url.URL, inputReader io.Reader, options ...
 		previousHashChannel <- EncodedHashChainInitialValue
 
 		failed := false
-
 		for {
-
-			buffer := pool.Get(writeOptions.blockSize)
-			bytesRead, err := io.ReadFull(inputReader, buffer)
+			var bytesRead int
+			var buffer []byte
+			var err error
+
+			if(writeOptions.timeWindow != 0) {
+				window := time.Second* time.Duration(writeOptions.timeWindow)
+				bufferTemp := make([]byte, 1)
+				bytesReadTemp := 0
+				for start := time.Now(); time.Since(start) < window; {
+					bytesReadTemp, err = io.ReadAtLeast(inputReader, bufferTemp, 1)
```

Comment:
I see, just thinking out loud below, let me know if any of this is not making any sense 😄 thanks for bearing with me! 

In this case if the "(echo "hi"; sleep 60;)" was in a loop say happening 30 times, we would write a few bytes after each 60 second sleep, in wont wait for error or EOF or error as long as it can read that one byte and timer is unblocked (which would still be faster then waiting for all say 1024 bytes to be read before continuing but I can still see how that is not ideal). Embarrassingly I missed that sleep is blocking Read from returning (didnt use a long enough window, duh), looks like the "hi" bytes do come through but not fast enough before sleep happens, so then another Read is called and it will wait until sleep is complete. I was naive to think that the time on the for loop was making the ReadAtLeast exit or not happen, but I can totally see now how that was completely wrong. 

Aha, reading more about concurrency and go routines I can see more things for me to investigate. Thanks for pointing this out. What would happen with concurrency is that I could start writing the bytes that were read, even if the most current Read is being waited on. (and then making sure everything is handled appropriately when it does return)

Other notes: 
- I figured ReadAtLeast by 1 byte is not ideal, but I am struggling a bit with how to determine what size to select, unless the size of the read data as adjusted as we go through few time windows to try and pick the maximum block size that is going to work most efficiently given the speed of the trickle. I figured we can probably pick a smaller size at first the makes sense - maybe 128 or something like that? 🤷 ( and maybe increase it as we go, as as long as the # of bytes that are being read is exactly the blockSize we picked, and if it gets smaller, then adjust - but maybe that is more complicated then it needs to be)
- I guess would this mean that writeOption.blockSize , the default value that is, would be ignored and set based on time window or not necessarily?

Reply:
I guess now that I think about it more, it doesnt quite matter what writeOption.blockSize is - it can stay the current value. As time goes on, we grab what is available and go from there.

Reply:
From the docs, by convention, `Reader.Read()` will return early if data is available. Regarding `writeOption.blockSize`, we should flush as soon as either we have `writeOption.blockSize` bytes or the timer has elapsed and we have data.